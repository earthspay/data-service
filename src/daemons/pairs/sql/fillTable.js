const knex = require('knex');
const pg = knex({ client: 'pg' });

const selectVolumeEarthsFromPairsCTE = pg({ d: 'pairs_cte' })
  .select({ volume_earths: pg.raw(`case when d.amount_asset_id='EARTHS' then p.quote_volume / d.weighted_average_price when d.price_asset_id='EARTHS' then p.quote_volume * d.weighted_average_price end`) })
  .whereRaw(`(d.amount_asset_id=p.price_asset_id AND d.price_asset_id='EARTHS')`)
  .orWhereRaw(
    `(d.price_asset_id=p.price_asset_id AND d.amount_asset_id='EARTHS')`
  );

const selectExchanges = pg('txs_7')
  .select(['price_asset', 'amount_asset', 'amount', 'price', 'time_stamp'])
  .whereRaw(`time_stamp >= now() - interval '1 day'`)
  .orderBy('time_stamp', 'desc');

const selectPairsCTE = pg
  .with('pairs_cte', qp => {
    qp.select({
      amount_asset_id: 'amount_asset',
      price_asset_id: 'price_asset',
      last_price: pg.raw(
        '(array_agg(e.price * 10 ^(-8 - p_dec.decimals + a_dec.decimals) ORDER BY e.time_stamp DESC)::numeric[])[1]'
      ),
      first_price: pg.raw(
        '(array_agg(e.price * 10 ^(-8 - p_dec.decimals + a_dec.decimals) ORDER BY e.time_stamp)::numeric[])[1]'
      ),
      volume: pg.raw('sum(e.amount * 10 ^(-a_dec.decimals))'),
      quote_volume: pg.raw(
        'sum(e.amount * 10 ^(-a_dec.decimals) * e.price * 10 ^(-8 - p_dec.decimals + a_dec.decimals))'
      ),
      weighted_average_price: pg.raw(
        'sum(e.amount * 10 ^(-a_dec.decimals) * e.price * 10 ^(-8 - p_dec.decimals + a_dec.decimals))/ sum(e.amount * 10 ^(-a_dec.decimals))'
      ),
      volume_earths: pg.raw(
        "case when amount_asset = 'EARTHS' then sum(e.amount * 10 ^(-a_dec.decimals)) when price_asset = 'EARTHS' then sum(e.amount * 10 ^(-a_dec.decimals) * e.price * 10 ^(-8 - p_dec.decimals + a_dec.decimals)) end"
      ),
      high: pg.raw('max(e.price * 10 ^(-8 - p_dec.decimals + a_dec.decimals))'),
      low: pg.raw('min(e.price * 10 ^(-8 - p_dec.decimals + a_dec.decimals))'),
      txs_count: pg.raw(
        'count(e.price)'
      ),
    })
      .from({ e: selectExchanges.clone() })
      .innerJoin(
        { a_dec: 'asset_decimals' },
        'e.amount_asset',
        'a_dec.asset_id'
      )
      .innerJoin({ p_dec: 'asset_decimals' }, 'e.price_asset', 'p_dec.asset_id')
      .groupBy(['amount_asset', 'price_asset']);
  })
  .from({ p: 'pairs_cte' })
  .columns(
    'amount_asset_id',
    'price_asset_id',
    'first_price',
    'last_price',
    'volume',
    {
      volume_earths: pg.raw(
        `coalesce(volume_earths, (${selectVolumeEarthsFromPairsCTE.toString()}))`
      ),
    },
    'quote_volume',
    'high',
    'low',
    'weighted_average_price',
    'txs_count'
  );

module.exports = tableName =>
  pg
    .into(tableName)
    .insert(selectPairsCTE)
    .toString();
