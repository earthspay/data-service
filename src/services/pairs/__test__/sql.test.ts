import { get, mget, search } from '../sql';

describe('sql query from pairs', () => {
  it('should get one pair', () => {
    expect(
      get({
        amountAsset: '111',
        priceAsset: '222',
      })
    ).toMatchSnapshot();
  });

  it('should get many pairs', () => {
    expect(
      mget([
        {
          amountAsset: '111',
          priceAsset: '222',
        },
        {
          amountAsset: '333',
          priceAsset: '444',
        },
      ])
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset', () => {
    expect(
      search({
        search_by_asset: '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj',
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset exactly', () => {
    expect(
      search({
        search_by_asset: '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj',
        match_exactly: true,
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset exactly', () => {
    expect(
      search({
        search_by_asset: '¯\\_(ツ)_/¯',
        match_exactly: true,
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for two assets (amount and price)', () => {
    expect(
      search({ search_by_assets: ['BTC', 'EARTHS'], limit: 10 })
    ).toMatchSnapshot();
  });

  it('should search pairs for two assets (amount and price)', () => {
    expect(
      search({ search_by_assets: ['¯\\_(ツ)_/¯', 'EARTHS'], limit: 10 })
    ).toMatchSnapshot();
  });

  it('should search pairs for two assets (amount and price)', () => {
    expect(
      search({
        search_by_assets: ['¯\\_(ツ)_/¯', 'EARTHS'],
        match_exactly: [true, false],
        limit: 10,
      })
    ).toMatchSnapshot();
  });
});
