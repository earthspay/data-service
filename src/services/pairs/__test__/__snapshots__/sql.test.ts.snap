// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql query from pairs should get many pairs 1`] = `"select \\"amount_asset_id\\", \\"price_asset_id\\", \\"first_price\\", \\"last_price\\", \\"volume\\", \\"quote_volume\\", \\"high\\", \\"low\\", \\"weighted_average_price\\", \\"txs_count\\", \\"volume_earths\\" from \\"pairs\\" as \\"t\\" where (\\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\") in (('111', '222'), ('333', '444'))"`;

exports[`sql query from pairs should get one pair 1`] = `"select \\"amount_asset_id\\", \\"price_asset_id\\", \\"first_price\\", \\"last_price\\", \\"volume\\", \\"quote_volume\\", \\"high\\", \\"low\\", \\"weighted_average_price\\", \\"txs_count\\", \\"volume_earths\\" from \\"pairs\\" as \\"t\\" where (\\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\") in (('111', '222'))"`;

exports[`sql query from pairs should search pairs for one asset 1`] = `"with \\"assets_cte\\" as (select \\"t\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or \\"t\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' or \\"t\\".\\"ticker\\" like '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' or to_tsvector(t.asset_name) @@ to_tsquery('7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj:*') union all select \\"t2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"t2\\" where \\"t2\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%') select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_earths\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" order by volume_earths desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for one asset exactly 1`] = `"with \\"assets_cte\\" as (select \\"t\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or \\"t\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' or \\"t\\".\\"ticker\\" like '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' or to_tsvector(t.asset_name) @@ to_tsquery('7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj:*') union all select \\"t2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"t2\\" where \\"t2\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%') select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_earths\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" order by volume_earths desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for one asset exactly 2`] = `"with \\"assets_cte\\" as (select \\"t\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = E'¯\\\\\\\\_(ツ)_/¯' or \\"t\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯%' or \\"t\\".\\"ticker\\" like E'¯\\\\\\\\_(ツ)_/¯%' or to_tsvector(t.asset_name) @@ to_tsquery('¯\\\\_ツ_/¯:*') union all select \\"t2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"t2\\" where \\"t2\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯%') select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_earths\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_id\\" = \\"t\\".\\"amount_asset_id\\" order by volume_earths desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for two assets (amount and price) 1`] = `"with \\"assets_cte\\" as (select \\"t1\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"t2\\".\\"asset_id\\" as \\"price_asset_id\\" from (select \\"a\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"a\\" where \\"a\\".\\"asset_id\\" = 'BTC' or \\"a\\".\\"asset_name\\" ilike 'BTC%' or \\"a\\".\\"ticker\\" like 'BTC%' or to_tsvector(a.asset_name) @@ to_tsquery('BTC:*') union all select \\"a2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"a2\\" where \\"a2\\".\\"asset_name\\" ilike 'BTC%') as \\"t1\\" cross join (select \\"p\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"p\\" where \\"p\\".\\"asset_id\\" = 'EARTHS' or \\"p\\".\\"asset_name\\" ilike 'EARTHS%' or \\"p\\".\\"ticker\\" like 'EARTHS%' or to_tsvector(p.asset_name) @@ to_tsquery('EARTHS:*') union all select \\"p2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"p2\\" where \\"p2\\".\\"asset_name\\" ilike 'EARTHS%') as \\"t2\\") select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_earths\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"direct_cte\\" on \\"direct_cte\\".\\"amount_asset_id\\" = \\"t\\".\\"amount_asset_id\\" and \\"direct_cte\\".\\"price_asset_id\\" = \\"t\\".\\"price_asset_id\\" union all select \\"t1\\".\\"amount_asset_id\\", \\"t1\\".\\"price_asset_id\\", \\"t1\\".\\"first_price\\", \\"t1\\".\\"last_price\\", \\"t1\\".\\"volume\\", \\"t1\\".\\"quote_volume\\", \\"t1\\".\\"high\\", \\"t1\\".\\"low\\", \\"t1\\".\\"weighted_average_price\\", \\"t1\\".\\"txs_count\\", \\"t1\\".\\"volume_earths\\" from \\"pairs\\" as \\"t1\\" inner join \\"assets_cte\\" as \\"reverse_cte\\" on \\"reverse_cte\\".\\"amount_asset_id\\" = \\"t1\\".\\"price_asset_id\\" and \\"reverse_cte\\".\\"price_asset_id\\" = \\"t1\\".\\"amount_asset_id\\" order by volume_earths desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for two assets (amount and price) 2`] = `"with \\"assets_cte\\" as (select \\"t1\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"t2\\".\\"asset_id\\" as \\"price_asset_id\\" from (select \\"a\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"a\\" where \\"a\\".\\"asset_id\\" = E'¯\\\\\\\\_(ツ)_/¯' or \\"a\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯%' or \\"a\\".\\"ticker\\" like E'¯\\\\\\\\_(ツ)_/¯%' or to_tsvector(a.asset_name) @@ to_tsquery('¯\\\\_ツ_/¯:*') union all select \\"a2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"a2\\" where \\"a2\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯%') as \\"t1\\" cross join (select \\"p\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"p\\" where \\"p\\".\\"asset_id\\" = 'EARTHS' or \\"p\\".\\"asset_name\\" ilike 'EARTHS%' or \\"p\\".\\"ticker\\" like 'EARTHS%' or to_tsvector(p.asset_name) @@ to_tsquery('EARTHS:*') union all select \\"p2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"p2\\" where \\"p2\\".\\"asset_name\\" ilike 'EARTHS%') as \\"t2\\") select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_earths\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"direct_cte\\" on \\"direct_cte\\".\\"amount_asset_id\\" = \\"t\\".\\"amount_asset_id\\" and \\"direct_cte\\".\\"price_asset_id\\" = \\"t\\".\\"price_asset_id\\" union all select \\"t1\\".\\"amount_asset_id\\", \\"t1\\".\\"price_asset_id\\", \\"t1\\".\\"first_price\\", \\"t1\\".\\"last_price\\", \\"t1\\".\\"volume\\", \\"t1\\".\\"quote_volume\\", \\"t1\\".\\"high\\", \\"t1\\".\\"low\\", \\"t1\\".\\"weighted_average_price\\", \\"t1\\".\\"txs_count\\", \\"t1\\".\\"volume_earths\\" from \\"pairs\\" as \\"t1\\" inner join \\"assets_cte\\" as \\"reverse_cte\\" on \\"reverse_cte\\".\\"amount_asset_id\\" = \\"t1\\".\\"price_asset_id\\" and \\"reverse_cte\\".\\"price_asset_id\\" = \\"t1\\".\\"amount_asset_id\\" order by volume_earths desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for two assets (amount and price) 3`] = `"with \\"assets_cte\\" as (select \\"t1\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"t2\\".\\"asset_id\\" as \\"price_asset_id\\" from (select \\"a\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"a\\" where \\"a\\".\\"asset_id\\" = E'¯\\\\\\\\_(ツ)_/¯' or \\"a\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯' or \\"a\\".\\"ticker\\" like E'¯\\\\\\\\_(ツ)_/¯' or to_tsvector(a.asset_name) @@ to_tsquery('¯\\\\_ツ_/¯') union all select \\"a2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"a2\\" where \\"a2\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯') as \\"t1\\" cross join (select \\"p\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets\\" as \\"p\\" where \\"p\\".\\"asset_id\\" = 'EARTHS' or \\"p\\".\\"asset_name\\" ilike 'EARTHS%' or \\"p\\".\\"ticker\\" like 'EARTHS%' or to_tsvector(p.asset_name) @@ to_tsquery('EARTHS:*') union all select \\"p2\\".\\"asset_id\\" as \\"asset_id\\" from \\"assets_metadata\\" as \\"p2\\" where \\"p2\\".\\"asset_name\\" ilike 'EARTHS%') as \\"t2\\") select \\"t\\".\\"amount_asset_id\\", \\"t\\".\\"price_asset_id\\", \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_earths\\" from \\"pairs\\" as \\"t\\" inner join \\"assets_cte\\" as \\"direct_cte\\" on \\"direct_cte\\".\\"amount_asset_id\\" = \\"t\\".\\"amount_asset_id\\" and \\"direct_cte\\".\\"price_asset_id\\" = \\"t\\".\\"price_asset_id\\" union all select \\"t1\\".\\"amount_asset_id\\", \\"t1\\".\\"price_asset_id\\", \\"t1\\".\\"first_price\\", \\"t1\\".\\"last_price\\", \\"t1\\".\\"volume\\", \\"t1\\".\\"quote_volume\\", \\"t1\\".\\"high\\", \\"t1\\".\\"low\\", \\"t1\\".\\"weighted_average_price\\", \\"t1\\".\\"txs_count\\", \\"t1\\".\\"volume_earths\\" from \\"pairs\\" as \\"t1\\" inner join \\"assets_cte\\" as \\"reverse_cte\\" on \\"reverse_cte\\".\\"amount_asset_id\\" = \\"t1\\".\\"price_asset_id\\" and \\"reverse_cte\\".\\"price_asset_id\\" = \\"t1\\".\\"amount_asset_id\\" order by volume_earths desc NULLS LAST limit 10"`;
